File: src/main/java/ninekothecat/catconomy/logging/CatLogger.java
package ninekothecat.catconomy.logging;

import ninekothecat.catconomy.Catconomy;
import ninekothecat.catconomy.enums.TransactionResult;
import ninekothecat.catconomy.interfaces.ICatLogger;
import ninekothecat.catconomy.interfaces.ITransaction;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.Marker;
import org.apache.logging.log4j.MarkerManager;
import org.apache.logging.log4j.core.LoggerContext;
import org.w3c.dom.DOMConfiguration;

import java.io.File;
import java.text.MessageFormat;
import java.util.Arrays;
import java.util.Objects;

public class CatLogger implements ICatLogger {
    private Logger logger;

    public CatLogger() {
        try {
            LoggerContext context = new LoggerContext("CCContext");

            context.setConfigLocation(Objects.requireNonNull(Catconomy.class.getClassLoader().getResource("loggerSettings/log4j2.xml")).toURI());
            logger = context.getLogger("CCTransactionLogger");
        }catch (Exception e){
            e.printStackTrace();
        }



    }

    @Override
    public void success(ITransaction transaction) {
        Marker marker = new MarkerManager.Log4jMarker(
                String.format("%s V%s", transaction.getPlugin().getName(),
                        transaction.getPlugin().getDescription().getVersion()));
        logger.info(marker, transaction.getMessage());
    }

    @Override
    public void fail(ITransaction transaction, TransactionResult result) {
        Marker marker = new MarkerManager.Log4jMarker(
                String.format("%s V%s", transaction.getPlugin().getName(),
                        transaction.getPlugin().getDescription().getVersion()));
        logger.warn(marker,
                MessageFormat.format("Transaction with type {0} Failed because {1} with amount {2} initiated by {3} with users {4}",
                transaction.getTransactionType().toString(), result.toString(), transaction.getAmount(), transaction.getInitiator(),
                Arrays.toString(transaction.getUsersInvolved().toArray())));
    }
    public void error(ITransaction transaction,TransactionResult result, Exception exception){
        Marker marker = new MarkerManager.Log4jMarker(
                String.format("%s V%s", transaction.getPlugin().getName(),
                        transaction.getPlugin().getDescription().getVersion()));
        logger.error(marker,
                MessageFormat.format("Transaction with type {0} failed because {1} with amount {2} initiated by {3} with users {4}",
                        transaction.getTransactionType().toString(),
                        result.toString(),
                        transaction.getAmount(),
                        transaction.getInitiator(),
                        Arrays.toString(transaction.getUsersInvolved().toArray())
                        ),exception);

    }
}

File: src/main/kotlin/ninekothecat/catconomy/logging/CatLogger.kt
package ninekothecat.catconomy.logging

import ninekothecat.catconomy.Catconomy
import ninekothecat.catconomy.interfaces.ICatLogger
import ninekothecat.catplugincore.money.enums.TransactionResult
import ninekothecat.catplugincore.money.interfaces.ITransaction
import org.apache.logging.log4j.Logger
import org.apache.logging.log4j.Marker
import org.apache.logging.log4j.MarkerManager.Log4jMarker
import org.apache.logging.log4j.core.LoggerContext
import org.bukkit.plugin.Plugin
import java.util.*

class CatLogger : ICatLogger {
    private var logger: Logger? = null
    override fun success(transaction: ITransaction) {
        val marker: Marker = generateLog4J2Marker(transaction.plugin)
        logger!!.info(marker, transaction.message)
    }

    override fun fail(transaction: ITransaction, result: TransactionResult) {
        val marker: Marker = generateLog4J2Marker(transaction.plugin)
        logger!!.warn(
            marker,
            "Transaction with type ${transaction.transactionType} Failed because $result with " +
                    "amount ${transaction.amount} initiated by ${transaction.initiator} " +
                    "with users ${transaction.usersInvolved!!.toTypedArray().contentToString()}"
        )
    }

    override fun error(transaction: ITransaction, result: TransactionResult, exception: Exception?) {
        val marker: Marker = generateLog4J2Marker(transaction.plugin)
        logger!!.error(
            marker,
            "Transaction with type ${transaction.transactionType} failed because $result with " +
                    "amount ${transaction.amount} initiated by ${transaction.initiator} " +
                    "with users ${transaction.usersInvolved!!.toTypedArray().contentToString()}", exception
        )
    }

    private fun generateLog4J2Marker(plugin: Plugin): Log4jMarker =
        Log4jMarker("${plugin.name} V${plugin.description.version}")

    init {
        try {
            val context = LoggerContext("CCContext")
            context.configLocation =
                Objects.requireNonNull(Catconomy::class.java.classLoader.getResource("loggerSettings/log4j2.xml"))
                    .toURI()
            logger = context.getLogger("CCTransactionLogger")
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}